src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs

```diff
@@ -88,25 +94,36 @@ public abstract class RegexRunner
 
-        protected internal Match? Scan(Regex regex, string text, int textbeg, int textend, int textstart, int prevlen, bool quick, TimeSpan timeout)
+        protected internal virtual void Scan(ReadOnlySpan<char> text)
         {
-            this.quick = quick;
+            string? s = runtext;
+            // The passed in span is sliced from runtextbeg to runtextend already, but in the precompiled scenario
+            // we require to use the complete input and to use the full string instead. We first test to ensure that the
+            // passed in span matches the original input by using the original runtextbeg. If that is not the case,
+            // then it means the user is calling the new span-based APIs using CompiledToAssembly, so we throw NSE
+            // so as to prevent a lot of unexpected allocations.
+            if (s == null || text != s.AsSpan(originalRuntextbeg, text.Length))
+            {
+                // If we landed here then we are dealing with a CompiledToAssembly case where the new Span overloads are being called.
+                throw new NotSupportedException(SR.UsingSpanAPIsWithCompiledToAssembly);
+            }
 
-            // Handle timeout argument
-            _timeout = -1; // (int)Regex.InfiniteMatchTimeout.TotalMilliseconds
-            bool ignoreTimeout = _ignoreTimeout = Regex.InfiniteMatchTimeout == timeout;
-            if (!ignoreTimeout)
+            // If the original beginning wasn't zero, then we have to adjust some of the
+            // internal fields of RegexRunner to ensure the Precompiled Go and FFC methods
+            // will continue to work as expected since they work over the original input, as opposed
+            // to using the sliced span.
+            if (originalRuntextbeg != 0)
             {
-                // We are using Environment.TickCount and not Stopwatch for performance reasons.
-                // Environment.TickCount is an int that cycles. We intentionally let timeoutOccursAt
-                // overflow it will still stay ahead of Environment.TickCount for comparisons made
-                // in DoCheckTimeout().
-                Regex.ValidateMatchTimeout(timeout); // validate timeout as this could be called from user code due to being protected
-                _timeout = (int)(timeout.TotalMilliseconds + 0.5); // Round;
-                _timeoutOccursAt = Environment.TickCount + _timeout;
-                _timeoutChecksToSkip = TimeoutCheckFrequency;
+                runtextbeg = originalRuntextbeg;
+                runtextstart += originalRuntextbeg;
+                runtextend += originalRuntextbeg;
             }
 
+            Scan(runregex!, s, originalRuntextbeg, originalRuntextbeg + text.Length, runtextstart + originalRuntextbeg, -1, quick, runregex!.internalMatchTimeout);
+        }
+
+        protected internal Match? Scan(Regex regex, string text, int textbeg, int textend, int textstart, int prevlen, bool quick, TimeSpan timeout)
```

Comment:
Don't these changes break the behavior of this protected method? e.g. it's no longer setting the fields based on these values, it's no longer validating the timeout, etc.?

Reply:
It does, if somebody was manually overriding this Scan method (which AFAIK the answer is nobody) they could end up setting the timeout and all of the rest of the fields twice, as tehy would have been set already via InitializeForScan and they would be resetting them.

Reply:
This method isn't virtual.  I'm not asking about someone overriding it, but about someone calling it.

Reply:
Oh right, sorry misunderstood your comment. Yes you are right, this is for sure breaking people directly calling this with values since pretty much all of the parameters now are thrown away essentially. If you think that there are potential consumers calling this, I could just replay the logic of re-setting the values inside this method body (with the added cost that there would be a small perf hit on CompiledToAssembly scenarios).

Reply:
I expect there's close to 0 users of this, but as it stands they'd all be broken, so we should fix it unless we explicitly choose to break it, in which case we should obsolete it and make it throw PNSE. At the same time, we don't _need_ to call this method ourselves, so if you're concerned about overhead, you can call to an internal method that this also uses. It's not virtual so there are no concerns about using an override.